### ISSUE TITLE
[RLLIB] Connector Pipeline is very slow

### ISSUE THREAD

**Issue by brieyla1** [ray-project/ray#50011]
### What happened + What you expected to happen

Hey !

I was Investigating RLLIB's pipelines and connectors, a nice dx upgrade from the previous messier and more hardcoded implementation of rllib.

Through my tests, I've noticed that most of the time spent in the `learning_step` was done in the `pipeline` rather than in the `model`, with only **19%** of the time being in the `model` class doing machine learning.

Around **70%** of the time is taken by the pipeline V2 -- **30%** of which just add the time dimension and zero pad the data, and **25%** adding cols to train batch. (see flamegraph below)

I have a feeling that this could be sped up by allowing the `space_utils`'s `batch` method to work with `tensors` directly when possible rather than `numpy`ing the lists every time.

```python
# https://github.com/ray-project/ray/blob/master/rllib/utils/spaces/space_utils.py#L325C1-L375C1

@DeveloperAPI
def batch(
    list_of_structs: List[Any],
    *,
    individual_items_already_have_batch_dim: Union[bool, str] = False,
):
    ....
    if individual_items_already_have_batch_dim == "auto":
        flat = tree.flatten(list_of_structs[0])
        individual_items_already_have_batch_dim = isinstance(flat[0], BatchedNdArray)

    np_func = np.concatenate if individual_items_already_have_batch_dim else np.stack
    ret = tree.map_structure(lambda *s: np_func(s, axis=0), *list_of_structs)
    return ret
```

If we go down the rabbit hole of checking the time flamegraph consumption of every item of the pipeline, the biggest consumer and culprit is generally a `map_structure` call. Keeping everything a `tensor` (and keeping on device is GPU / CPU) when possible would drastically speed up these processes and avoid unnecessary `GPU -> CPU -> GPU` calls that seem to bottleneck the performance.

![Image](https://github.com/user-attachments/assets/d4a05b51-c4e0-4d7c-84e3-4ab613f0fe03)

I'm going to try and optimize a few of these, I wonder if @sven1977 the wizard has any insight on this. Updating this could theoritically make a lot of the enterprise hardware get a free **30-40%** speed increase with no compromise on RLLIB new api.

Nota bene: Some optimizations could also be made to the `env_runner` connectors too, as something similar is happening. where a big portion of the time is spent creating the action distribution class in python rather than computing the logits and the action itself. _**But this is another problem for another day.**_

### Versions / Dependencies

GH200 machine | 64 GRACE CPU cores (arm), 463.9 GB RAM, 4.4 TB SSD, H100

Ubuntu 22.04.5 LTS
Python 3.12.2
Ray @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-3.0.0.dev0-cp312-cp312-manylinux2014_aarch64.whl#sha256=9131183c34d496b7b3f3034c227f70e66891653a0d2313c6fc1d7376bd7a6e66

### Reproduction script

batch_size=256K steps 
minibatch=8K steps
num_epochs=4 epochs
4 env runners each 256 envs

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by brieyla1** [comment]
Hey all !

Although it may not be as good as the suggested implementation of keeping the data as tensors & such. I appreciate the amount of work that needs to be done to work this out. Hence, I've create a quick plug-n-play speedup solution that improves the speed of the bad connectors without changing the data output. Nothing changes besides less time spent in the connectors. https://github.com/ray-project/ray/pull/50035

NOTE: Performance in the connectors is still horrendous; but it went from 70% to 40% (see flamegraph below) This is essentially a free speed-up.


![Image](https://github.com/user-attachments/assets/4630c6b5-fc05-4aa9-8209-30a10fdcb5b2)

**Comment by brieyla1** [comment]
#50035

**Comment by sven1977** [comment]
Hey @brieyla1 , thanks so much for raising this issue! We are observing the same thing in some workloads on our end.
I think we already have a fix for this, which is not activated by default, though. Could you try setting in your config:

```
config.env_runners(episodes_to_numpy=False)
```

and let us know, whether this makes any difference for you?

If yes, this'll give us another hint that we should switch the default behavior or even get rid of the numpy'zation altogether.

**Comment by sven1977** [comment]
Note that this is a new setting that probably only exists on the master branch (should be part of 2.42, which is released in 1-2 weeks).

**Comment by brieyla1** [comment]
Hey Sven, thanks for the response!

I've taken a look at the performance with the `episodes_to_numpy=False`. The idea seems promising, however, the performance is terrible, going from **+-30s** per iter in my benchmark notebook to **+-60s** with it enabled on on the latest ray/master branch (**2x** slower). My fixes at https://github.com/ray-project/ray/pull/50035 make the performance to **+-24s** but keeps numpy'zation which makes me think there's a better solution. It's stable in my scenario but I'd have to run more tests to ensure it's stable all over the repo.

We could connect on Slack / Discord and work on a more permanent fix. We have cluster training and optimizing costs is one of the priorities on our side.

**Comment by sven1977** [comment]
Hey @brieyla1 , thanks for the quick response. I wonder why your data is already in tensor format before(!) it goes into these (slow) connector pieces, like the zero-padder.

You are saying:
```
I have a feeling that this could be sped up by allowing the space_utils's batch method to work with tensors directly when possible rather than numpying the lists every time.
```
The `batch` method should run on numpy arrays and only the `NumpyToTensor` connector piece (at the very end of the pipeline) should translate to torch. Do you have a custom connector that already creates torch tensors? Yes, maybe this is the reason, why the pipeline is slow in your case (back and forth copying between CPU and GPU). 🤔

**Comment by brieyla1** [comment]
Hey Sven,

I haven't got any custom connector in my setup; maybe the problem isn't about CPU <-> GPU calls but simply the number of connector operations and the lack of optimizations & single-core CPU-bounded computations.

Creating an optimized default connector rather than multiple ones would make it a lot faster and decrease the O() complexity of the connector logic (there are a lot of tree.map() fn, which can go explosive for large amounts of data & batch size)

Would you like me to create a reproduction script?

thanks

**Comment by LLudoDev** [comment]
[fast_add_columns_connector.py](https://github.com/user-attachments/files/26383050/fast_add_columns_connector.py)
For those who still have the issue: using this connector in your config will save you a LOT of training time. 👍
